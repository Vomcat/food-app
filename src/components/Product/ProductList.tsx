import { useEffect, useState } from "react";

import styled from "styled-components";
import { respondFrom, breakpoints, dimensions } from "styles";

import { ProductItemProps } from "interfaces/Product";

import ProductSortBar from "components/Product/ProductSortBar";
import Item from "components/Product/ProductItem";

const ProductListWrapper = styled.div`
  display: grid;
  align-items: center;
  justify-content: center;
  grid-template-columns: minmax(215px, 320px);
  gap: ${dimensions.spacing.lg}px;
  padding: 50px 0;

  ${respondFrom(breakpoints.tablet)`
        grid-template-columns: repeat(2,minmax(215px, 320px));
        gap: ${dimensions.spacing.lg3}px ${dimensions.spacing.lgmax}px;
    `}

  ${respondFrom(breakpoints.desktop)`
        grid-template-columns: repeat(3,minmax(215px, 1fr));
        gap: ${dimensions.spacing.lg3}px ${dimensions.spacing.lgmax}px;
        padding: 100px 0;
    `}
`;

const ProductList = () => {
  const [items, setItems] = useState<ProductItemProps[]>([]);

  const fetchData = async () => {
    try {
      const response = await fetch(" http://localhost:3000/items");
      const data = await response.json();
      setItems(data);
    } catch (err) {
      throw new Error("Something went wrong");
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  const onSelectSortHandler = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const sortDirection = e.target.value;
    const copyArray = [...items];

    copyArray.sort((a: ProductItemProps, b: ProductItemProps) => {
      return sortDirection === "0"
        ? a.id - b.id
        : sortDirection === "1"
        ? a.name.localeCompare(b.name)
        : sortDirection === "2"
        ? a.price - b.price
        : b.price - a.price;
    });

    setItems(copyArray);
  };

  return (
    <>
      <ProductSortBar changeHandler={onSelectSortHandler} />
      <ProductListWrapper>
        {items.map(({ id, name, price, imageURI }) => (
          <Item
            key={id}
            id={id}
            name={name}
            price={price}
            imageURI={imageURI}
          />
        ))}
      </ProductListWrapper>
    </>
  );
};

export default ProductList;
