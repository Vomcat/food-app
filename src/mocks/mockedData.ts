export const MOCK_DATA = [
  {
    id: 1,
    image: "/assets/images/item1.png",
    name: "Mniam Mniam Food",
    price: 5,
  },
  {
    id: 2,
    image: "/assets/images/item2.png",
    name: "Mniam Mniam Food as adsdasdasd",
    price: 5,
  },
];
