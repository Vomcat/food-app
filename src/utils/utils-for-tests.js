import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import cartSlice from "store/cartSlice";

const renderWithProviders = function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { cartReducers: cartSlice },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export { renderWithProviders as render };
export * from "@testing-library/react";
