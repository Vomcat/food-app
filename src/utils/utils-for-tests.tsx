import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState, Store } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { BrowserRouter } from "react-router-dom";

import cartSlice from "store/cartSlice";
import type { RootState } from "store/store";

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions
  extends Omit<RenderOptions, "wrapper" | "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: Store;
}

const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = { items: [], totalAmount: 0, totalQuantity: 0 },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { cartReducers: cartSlice },
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export { renderWithProviders as render };
export * from "@testing-library/react";
export * from '@testing-library/jest-dom';

