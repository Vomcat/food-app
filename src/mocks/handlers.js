import { rest } from "msw";
import { MOCK_DATA_PRODUCTS, MOCK_DATA_ORDERS } from "mocks/mockedData";

export const handlers = [
  rest.get(`${process.env.REACT_APP_API_HOST}/items`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(MOCK_DATA_PRODUCTS));
  }),

  rest.get(`${process.env.REACT_APP_API_HOST}/order`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(MOCK_DATA_ORDERS));
  }),
];
