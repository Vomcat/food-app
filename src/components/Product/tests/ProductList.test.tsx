import { render, screen } from "utils/utils-for-tests";

import ProductList from "components/Product/ProductList";

import { MOCK_DATA_PRODUCTS } from "mocks/mockedData";

describe("Product List render test", () => {
  test("Products list renders correctly", async () => {
    const { asFragment } = render(<ProductList />);

    await screen.findByText(MOCK_DATA_PRODUCTS[0].name);

    expect(asFragment()).toMatchSnapshot();
  });

  test("Products list renders correctly", async () => {
     render(<ProductList />);

  });
});
