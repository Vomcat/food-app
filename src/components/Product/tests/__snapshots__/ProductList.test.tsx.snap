// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component Products list renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe SGnSj"
  >
    <select
      class="sc-gswNZR hnRSQm"
      name="Sort by"
    >
      <option
        selected=""
        value="0"
      >
        Default sorting
      </option>
      <option
        value="1"
      >
        Alphabetical
      </option>
      <option
        value="2"
      >
        Lowest price
      </option>
      <option
        value="3"
      >
        Highest price
      </option>
    </select>
  </div>
  <div
    class="sc-ftTHYK iKvzqA"
  >
    <div
      class="sc-eDvSVe gIeMic"
    >
      <div
        class="sc-jSUZER"
      >
        <img
          alt="Mniam Mniam Food"
        />
      </div>
      <p
        class="sc-gKPRtg csZcBY"
      >
        Mniam Mniam Food
      </p>
      <div
        class="sc-iBYQkv kzNpzW"
      >
        <p
          class="sc-gKPRtg csZcBY"
        >
          5 $
        </p>
        <button
          class="sc-hLBbgP iGGqbO"
          type="button"
        >
          <img
            alt="plus icon"
            src="plus.png"
          />
        </button>
      </div>
    </div>
    <div
      class="sc-eDvSVe gIeMic"
    >
      <div
        class="sc-jSUZER"
      >
        <img
          alt="Mniam Mniam Food as adsdasdasd"
        />
      </div>
      <p
        class="sc-gKPRtg csZcBY"
      >
        Mniam Mniam Food as adsdasdasd
      </p>
      <div
        class="sc-iBYQkv kzNpzW"
      >
        <p
          class="sc-gKPRtg csZcBY"
        >
          5 $
        </p>
        <button
          class="sc-hLBbgP iGGqbO"
          type="button"
        >
          <img
            alt="plus icon"
            src="plus.png"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
